src/query_graph/graph_path.rs

```diff
@@ -238,4 +481,770 @@ impl OpGraphPath {
+        //  -  `self`: <some prefix> -[t]-> T(A)               -[x] -> Int(A)
+        //  - `other`: <some prefix> -[t]-> T(A) -[key]-> T(B) -[x] -> Int(B)
+        // because when the code is asked for the options for `x` after `<some prefix> -[t]-> T(A)`,
+        // it notices that `x` is a leaf and is in `A`, so it doesn't ever look for alternative
+        // paths. But this only works for direct leaves of an entity. In the example at the start,
+        // field `u` makes this not work, because when we compute choices for `u`, we don't yet know
+        // what comes after that, and so we have to take the option of going to subgraph `B` into
+        // account (it may very well be that whatever comes after `u` is not in `A`, for instance).
+        let self_tail_weight = self.graph.node_weight(self.tail)?;
+        let other_tail_weight = self.graph.node_weight(other.tail)?;
+        if self_tail_weight.source == other_tail_weight.source {
+            // As described above, we want to know if one of the paths has no jumps at all (after
+            // the common prefix) while the other has some.
+            self.compare_subgraph_jumps_after_last_common_node(other)
+        } else {
+            Ok(Ordering::Equal)
+        }
+    }
+}
+
+impl Display for OpGraphPath {
+    fn fmt(&self, _f: &mut Formatter<'_>) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+impl SimultaneousPaths {
+    /// Given options generated for the advancement of each path of a `SimultaneousPaths`, generate
+    /// the options for the `SimultaneousPaths` as a whole.
+    fn flat_cartesian_product(
+        options_for_each_path: Vec<Vec<SimultaneousPaths>>,
+    ) -> Vec<SimultaneousPaths> {
+        // This can be written more tersely with a bunch of `reduce()`/`flat_map()`s and friends,
+        // but when interfaces type-explode into many implementations, this can end up with fairly
+        // large `Vec`s and be a bottleneck, and a more iterative version that pre-allocates `Vec`s
+        // is quite a bit faster.
+        if options_for_each_path.is_empty() {
+            return vec![];
+        }
+
+        // Track, for each path, which option index we're at.
+        let mut option_indexes = std::iter::repeat_with(|| 0)
+            .take(options_for_each_path.len())
+            .collect::<Vec<usize>>();
+
+        // Pre-allocate `Vec` for the result.
+        let num_options = options_for_each_path
+            .iter()
+            .fold(1, |acc, options| acc * options.len());
+        let mut product = Vec::with_capacity(num_options);
+
+        // Compute the cartesian product.
+        for _ in 0..num_options {
+            let num_simultaneous_paths = options_for_each_path
+                .iter()
+                .zip(option_indexes.iter())
+                .fold(0, |acc, (options, option_index)| {
+                    acc + options[*option_index].0.len()
+                });
```

Comment:
`zip` accepts `IntoIterator`, not just `Iterator`:

```suggestion
                .zip(&option_indexes)
                .map(|(options, option_index)| {
                    options[*option_index].0.len()
                })
                .sum();
```